okhttp/src/test/java/io/grpc/okhttp/OkHttpClientStreamTest.java

```diff
@@ -0,0 +1,151 @@
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.grpc.okhttp;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.verifyNoMoreInteractions;
+
+import io.grpc.Metadata;
+import io.grpc.MethodDescriptor.MethodType;
+import io.grpc.Status;
+import io.grpc.internal.ClientStreamListener;
+import io.grpc.okhttp.internal.framed.ErrorCode;
+import io.grpc.okhttp.internal.framed.Header;
+
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+import org.mockito.Mock;
+import org.mockito.Mockito;
+import org.mockito.MockitoAnnotations;
+import org.mockito.invocation.InvocationOnMock;
+import org.mockito.stubbing.Answer;
+
+import java.io.InputStream;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicReference;
+
+@RunWith(JUnit4.class)
+public class OkHttpClientStreamTest {
+  private static final int MAX_MESSAGE_SIZE = 100;
+
+  @Mock private AsyncFrameWriter frameWriter;
+  @Mock private OkHttpClientTransport transport;
+  @Mock private Runnable startCallback;
+  @Mock private OutboundFlowController flowController;
+  private final Object lock = new Object();
+  private final List<Header> requestHeaders = new ArrayList<Header>();
+
+  private OkHttpClientStream stream;
+
+  @Before
+  public void setUp() {
+    MockitoAnnotations.initMocks(this);
+    stream = new OkHttpClientStream(frameWriter, transport, startCallback, flowController,
+        MethodType.UNARY, lock, requestHeaders, MAX_MESSAGE_SIZE);
+  }
+
+  @Test
+  public void getType() {
+    assertEquals(MethodType.UNARY, stream.getType());
+  }
+
+  @Test
+  public void request() {
+    assertEquals(MethodType.UNARY, stream.getType());
+  }
+
+  @Test
+  public void sendCancel_notStarted() {
+    final AtomicReference<Status> statusRef = new AtomicReference<Status>();
+    stream.start(new BaseClientStreamListener() {
+      @Override
+      public void closed(Status status, Metadata trailers) {
+        statusRef.set(status);
+        assertTrue(Thread.holdsLock(lock));
+      }
+    });
+
+    stream.sendCancel(Status.CANCELLED);
+
+    assertEquals(Status.Code.CANCELLED, statusRef.get().getCode());
+  }
+
+  @Test
+  public void sendCancel_started() {
+    stream.start(new BaseClientStreamListener());
+    stream.start(1234);
```

Comment:
Yeah.... it really should have another name.  start(id)  is when concurrent streams have not been reached.